src/mysql.erl

```diff
@@ -269,14 +244,84 @@ query(Conn, Query, Params, FilterMap) when is_list(Params) andalso (is_function(
 %% The minimum time the prepared statement is cached can be specified using the
 %% option `{query_cache_time, Milliseconds}' to start_link/1.
 %%
-%% The return value is the same as for query/2.
+%% Results are returned in the form `{ok, ColumnNames, Rows}' if there is one
+%% result set. If there are more than one result sets, they are returned in the
+%% form `{ok, [{ColumnNames, Rows}, ...]}'.
+%%
+%% For queries that don't return any rows (INSERT, UPDATE, etc.) only the atom
+%% `ok' is returned.
+%%
+%% The `Params', `FilterMap' and `Timeout' arguments are optional.
+%% <ul>
+%%   <li>If the `Params' argument is the atom `no_params' or is omitted, a plain
+%%       query will be executed instead of a parameterized one.</li>
+%%   <li>If the `FilterMap' argument is the atom `no_filtermap_fun' or is
+%%       omitted, no row filtering/mapping will be applied and all result rows
+%%       will be returned unchanged.</li>
+%%   <li>If the `Timeout' argument is the atom `default_timeout' or is omitted,
+%%       the timeout given in `start_link/1' is used.</li>
+%% </ul>
+%%
+%% If the `FilterMap' argument is used, it must be a function of arity 1 or 2
+%% that returns either `true', `false', or `{true, Value}'.
+%%
+%% Each result row is handed to the given function as soon as it is received
+%% from the server, and only when the function has returned, the next row is
+%% fetched. This provides the ability to prevent memory exhaustion; on the
+%% other hand, it can cause the server to time out on sending if your function
+%% is doing something slow (see the MySQL documentation on `NET_WRITE_TIMEOUT').
+%%
+%% If the function is of arity 1, only the row is passed to it as the single
+%% argument, while if the function is of arity 2, the column names are passed
+%% in as the first argument and the row as the second.
+%%
+%% The value returned is then used to decide if the row is to be included in
+%% the result(s) returned from the `query' call (filtering), or if something
+%% else is to be included in the result instead (mapping). You may also use
+%% this function for side effects, like writing rows to disk or sending them
+%% to another process etc.
+%%
+%% Here is an example showing some of the things that are possible:
+%% ```
+%% Query = "SELECT `a`, `b`, `c` FROM `foo`",
```

Comment:
Very good documentation!

Only one thing: I don't like the backticks around column names in MySQL examples though (nor in actual programs) because it's not standard SQL and it's not needed. Just remove them, I'd prefer.

Regarding not having the full manual in the code, have you seen that there is an edoc file in the doc/ folder? Just saying. I don't mind that the full documentation about this is here in the code for now. We can move this to e.g. asciidocs some other time.

Reply:
> Only one thing: I don't like the backticks around column names in MySQL examples though (nor in actual programs) because it's not standard SQL and it's not needed. Just remove them, I'd prefer.

Sure, done :)

> Regarding not having the full manual in the code, have you seen that there is an edoc file in the doc/ folder? Just saying. I don't mind that the full documentation about this is here in the code for now. We can move this to e.g. asciidocs some other time.

Oh, no, I didn't notice. Looking at it, though, it seems that it would require some more structure and context, if you know what I mean. Like, just slapping the query paragraph from `mysql.erl` in there would look half-heartedly, ie not too good ;) I feel that there should be all things from connectin through querying in all sorts of ways etcetc in it, too. That's beyond my scope for now, though, so I'd say we leave the manual in the code for the time being until we can do an all-out manual ;)

Reply:
Good. I agree.